module/zfs/spa_misc.c

```diff
@@ -607,6 +617,30 @@ spa_config_held(spa_t *spa, int locks, krw_t rw)
  * ==========================================================================
  */
 
+spa_t *
+spa_lookup_lite(const char *name)
+{
+	static spa_t search;	/* spa_t is large; don't allocate on stack */
```

Comment:
Doesn't this burn the metaphorical house down on running multiple calls to it at once, or do I not understand how static declarations work in the kernel?

Reply:
I suppose, less obliquely.

What does a second lock buy you that just making spa_namespace_lock a rwlock doesn't? I get "it's too big to fix all at once", but if you just declare them all writers except for the codepaths you're only taking lite on here, it seems to produce the same outcomes, no?

And if not, I would predict similar confusion on people trying to consume this interface.

Reply:
> Doesn't this burn the metaphorical house down on running multiple calls to it at once

It's ok in this case since it's protected by spa_namespace_lock_lite.

> declare them all writers except for the codepaths you're only taking lite on here, it seems to produce the same outcomes, no?

Unfortunately, if one of the pools gets hosed up while holding spa_namespace_lock as writer, zpool status will hang trying to hold the reader lock.

Reply:
Isn't that still true on the lite lock, then?

Reply:
Now that you're making me think about, we may not actually need a separate `spa_namespace_lite_avl` tree.  We may just need the new lock.  The two separate trees make it conceptually easier to understand, but I don't know if it's functionally needed, since the trees must always be exactly the same.

Reply:
> Isn't that still true on the lite lock, then?

Technically yes, but realistically no.  The lite lock is only held as writer when adding/removing from the AVL tree (which is a very short, non-blocking operation):
```C
       rw_enter(&spa_namespace_lite_lock, RW_WRITER);                            
       avl_remove(&spa_namespace_avl, spa);             
       rw_exit(&spa_namespace_lite_lock);
...
       rw_enter(&spa_namespace_lite_lock, RW_WRITER);                            
       avl_add(&spa_namespace_avl, spa);  
       rw_exit(&spa_namespace_lite_lock);
```
All the other times the lite lock is taken as reader.

Reply:
I don't think the two separate trees save you anything, since I think the only time they'd be useful is if you wanted to access the ro tree while the rw tree had a lock held, and that isn't allowed with this dance anyway.

Reply:
I removed the lite tree and my local testing showed it worked.  The lite tree is removed in my latest push.